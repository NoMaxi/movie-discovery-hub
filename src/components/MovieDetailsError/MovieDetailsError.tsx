import { Link, useRouteError } from "react-router-dom";
import { CloseDetailsButton } from "@/components/common/CloseDetailsButton/CloseDetailsButton";

export const MovieDetailsError = () => {
    const error = useRouteError();
    console.error(error);

    return (
        <div className="error-container flex flex-col items-center justify-center relative h-full">
            <CloseDetailsButton />

            <p>Sorry, we couldn't load the movie details.</p>
            <p>Please try again later or return to search.</p>

            <Link to="/" className="btn mt-5">
                Back to Search
            </Link>
        </div>
    );
};
