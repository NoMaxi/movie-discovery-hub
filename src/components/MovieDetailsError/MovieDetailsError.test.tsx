import { render, screen } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import "@testing-library/jest-dom";
import { MovieDetailsError } from "./MovieDetailsError";

jest.mock("react-router-dom", () => {
    const actual = jest.requireActual("react-router-dom");
    return {
        ...actual,
        useRouteError: jest.fn(),
    };
});

jest.mock("@/components/common/CloseDetailsButton/CloseDetailsButton", () => ({
    CloseDetailsButton: () => <div data-testid="close-details-button" />,
}));

const { useRouteError } = jest.requireMock("react-router-dom");

describe("MovieDetailsError", () => {
    beforeEach(() => {
        jest.clearAllMocks();
        console.error = jest.fn();
    });

    const renderMovieDetailsError = (error = {}) => {
        useRouteError.mockReturnValue(error);
        return render(
            <MemoryRouter>
                <MovieDetailsError />
            </MemoryRouter>,
        );
    };

    test("Should render component with correct content", () => {
        const { asFragment } = renderMovieDetailsError();

        expect(screen.getByText("Sorry, we couldn't load the movie details.")).toBeInTheDocument();
        expect(screen.getByText("Please try again later or return to search.")).toBeInTheDocument();
        expect(screen.getByText("Back to Search")).toBeInTheDocument();
        expect(screen.getByTestId("close-details-button")).toBeInTheDocument();
        expect(asFragment()).toMatchSnapshot();
    });

    test("Should render a link back to the search page", () => {
        renderMovieDetailsError();

        const backLink = screen.getByTestId("back-to-search-link");
        expect(backLink).toBeInTheDocument();
        expect(backLink).toHaveAttribute("href", "/");
        expect(backLink).toHaveClass("btn mt-5");
        expect(backLink).toHaveTextContent("Back to Search");
    });

    test("Should log the error to console", () => {
        const testError = { message: "Test error message" };
        renderMovieDetailsError(testError);

        expect(console.error).toHaveBeenCalledWith(testError);
    });

    test("Should have correct container styling", () => {
        renderMovieDetailsError();

        const container = screen.getByTestId("error-container");
        expect(container).toHaveClass("error-container");
        expect(container).toHaveClass("flex");
        expect(container).toHaveClass("flex-col");
        expect(container).toHaveClass("items-center");
        expect(container).toHaveClass("justify-center");
        expect(container).toHaveClass("relative");
        expect(container).toHaveClass("h-full");
    });
});
