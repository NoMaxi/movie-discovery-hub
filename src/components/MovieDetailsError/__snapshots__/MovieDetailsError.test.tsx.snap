// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieDetailsError Should render component with correct content 1`] = `
<DocumentFragment>
  <div
    class="error-container flex flex-col items-center justify-center relative h-full"
    data-testid="error-container"
  >
    <div
      data-testid="close-details-button"
    />
    <p>
      Sorry, we couldn't load the movie details.
    </p>
    <p>
      Please try again later or return to search.
    </p>
    <a
      class="btn mt-5"
      data-discover="true"
      data-testid="back-to-search-link"
      href="/"
    >
      Back to Search
    </a>
  </div>
</DocumentFragment>
`;
